### DISCUSSION TITLE
Pyside and pyside@2 force use of Py_Limited_API, should be optional Opt-in instead

### DISCUSSION THREAD

**Discussion by dirkolbrich** [Homebrew/discussions#3810] [Writing Formulae/Casks]
The pyside and pyside@2 package force the use of the Py_Limited_API, which in turn breaks builds of software which uses this package but are not Py_Limited_API compatible.

See a rather lengthly discussion thread on the FreeCAD forum - [[Re: Mac build and python C API](https://forum.freecadweb.org/viewtopic.php?p=634260#p634260)](https://forum.freecadweb.org/viewtopic.php?p=634260#p634260)

The use of the Py_Limited_API was introduced with the following commits, yet no discussion can be found for this breaking change.

- Force limited API introduced for pyside6 with v6.3 in commit [8a37288](https://github.com/Homebrew/homebrew-core/commit/8a37288785664611bb02965e88a7a12156f70d33) at 2022-06-16

- Force limited API introduced for pyside@2 with v5.15.3 in commit [76135eb](https://github.com/Homebrew/homebrew-core/commit/76135eb8807e1721e647d9faa475a88ce128b0eb) at 2022-05-15

I do assume, this change to the pyside formulae was introduced as the Pyside project states in its [README](https://github.com/pyside/pyside-setup#development-options) that Py_Limited_API is set to yes by default.

BUT within the source code the Py_Limited_API is NOT set by default, it rather is an option which needs explicit manually enabling, see also the official [QT for Python cross compiling example](https://doc.qt.io/qtforpython-6/gettingstarted.html#cross-compiling-pyside). Here are the relevant parts of the code I could identify:
https://github.com/pyside/pyside-setup/blob/69d43ea814788a18e8d77920d493fd84f6fc699a/sources/shiboken6/CMakeLists.txt#L77-L84
https://github.com/pyside/pyside-setup/blob/69d43ea814788a18e8d77920d493fd84f6fc699a/build_scripts/main.py#L780-L788

Rather than force the use of the Py_Limited_API, this setting should be an Opt-in during install, not a default and unchangeable setting, e.g. install the package with `brew install pyside —with-limited-api`.

**Comment by SMillerDev** [comment]
We don't do options in Homebrew core, and since upstream enables it by default we will honour that default as we usually do.

**Reply by dirkolbrich** [reply]
~~That's the point. The upstream project does NOT enable Py_Limited_API by default.~~ Sorry, I read the source code wrong, Py_Limited_API is in fact set ON by default.

Edit: See comment by @Bo98, force enabling Py_Limited_API seems to be a workaround for a no longer existing bug.

**Comment by Bo98** [comment]
> BUT within the source code the Py_Limited_API is NOT set by default, it rather is an option which needs explicit manually enabling, see also the official [QT for Python cross compiling example](https://doc.qt.io/qtforpython-6/gettingstarted.html#cross-compiling-pyside). Here are the relevant parts of the code I could identify:
https://github.com/pyside/pyside-setup/blob/69d43ea814788a18e8d77920d493fd84f6fc699a/sources/shiboken6/CMakeLists.txt#L77-L84
https://github.com/pyside/pyside-setup/blob/69d43ea814788a18e8d77920d493fd84f6fc699a/build_scripts/main.py#L780-L788

The default "yes" implies it's meant to be enabled by default, and is documented in the code to behave like that. It doesn't behave like that due to a bug. (`option` converts "yes" to a boolean `ON` but other code expects a literal "yes" string.) Maybe I should open a bug report upstream and get that cleared up.

The reason for force enabling it was something else however: at the time of that commit `pyside@2` did _not_ support Python 3.10 without limited API enabled. This was reported upstream (https://bugreports.qt.io/browse/PYSIDE-1775) and the response was "we always considered in 5.15 that 'not limited api' means python 2.7 only", which is quite clear that Limited API should always be used on Python 3.

It looks like that pyside2 was fixed in the meantime. The buggy code was either fixed in 5.15.4 and 5.15.5.

As for pyside6, it probably just was made consistent - I wasn't involved with that change.

**Reply by dirkolbrich** [reply]
If the reason to force enable Py_Limited_API is no longer valid (due to fixing an upstream bug), shouldn't this "workaround" be removed from the formulae?

Edit: rereading the [README](https://github.com/pyside/pyside-setup#development-options) of the pyside source states the following:

>Development Options
>
>For development purposes the following options might be of use, when using setup.py build:
>[...]
>--limited-api=yes|no, default yes if applicable. Set or clear the limited API flag. Ignored for Python 2.

This implies IMHO that the Py_Limited_API option should only be set during development of the pyside project, not for the end user -> consumer of the `brew install pyside` package.

The problem is, that this setting within the brew pyside formulae now forces user of pyside on macOS to make their code Py_Limited_API conforming as well. We (the FreeCAD project) do not have this problem with Windows and Linux (nix) provided pyside packages.

**Reply by Bo98** [reply]
I wouldn't call it a workaround if upstream don't recommend anyone uses non-limited API on Python 3, but I'll raise an issue upstream in a moment to clarify that. The official distribution on PyPI uses limited API - which is a very clear indication that's the way they want to build things.

Removing limited API would that we can only support one version of Python at a time. I did plan to allow any Python 3.7-3.11 to be used but that won't be possible if limited API is disabled.

**Reply by Bo98** [reply]
Ok, I've heard back and the default was indeed bugged: https://bugreports.qt.io/browse/PYSIDE-2091.

In 6.4.1 and 6.2.7, the default will change so that Limited API is ON by default, as intended. I expect downstream packages will pick up that change when 6.4.1 is released in 3 weeks, excluding distributions that need to delay for API or ABI compatibility reasons.

**Reply by Bo98** [reply]
I do apologise for the disruption it caused you however - it was never intentional to break your build and these effects were not known at the time. Usually Limited API does not cause compatibility issues like this. Shiboken is somewhat an exception due to how much it exposes its internals in public headers and varies its API depending on whether Python Limited API is enabled, and by extension force sets `Py_LIMITED_API` in its public CMake helpers (and vice versa if we had it off).

Hopefully this alignment with upstream will clear things up going forward however.

**Reply by dirkolbrich** [reply]
Thank you for pursuing this topic further and keeping me posted.

Can I assume the line `-DFORCE_LIMITED_API=yes` in the homebrew pyside formulae can be removed for the next patch (whoever submits a PR for that), as it is set upstream by default? That was one of the reasons for my confusion, as this line suggested an explicit setting by the formulae but not on the upstream source.

Regarding the Shiboken force setting of `Py_Limited_API` in its public header it was a suspicion of our investigation as well. Yet we don't know, why it seems to only affect the build chain with the homebrew installed formulae in macOS, but not for builds in Linux and Windows.

We suspect, that this force setting of `Py_Limited_API` somehow spills over to consuming code of Shiboken, instead of just being applicable to the package itself. By thought, a package which compiles with `Py_Limited_API` enabled, should also work if this setting is no enabled. But this is a topic concerning only the interaction of our source and upstream Pyside/Shiboken, not Homebrew.
